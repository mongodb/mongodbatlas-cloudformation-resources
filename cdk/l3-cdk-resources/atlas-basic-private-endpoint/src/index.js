"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
exports.__esModule = true;
var privateEndpoint = require("@mongodbatlas-awscdk/private-endpoint");
var cdk = require("aws-cdk-lib");
var aws_cdk_lib_1 = require("aws-cdk-lib");
var AtlasBasePrivateEndpoint = /** @class */ (function (_super) {
    __extends(AtlasBasePrivateEndpoint, _super);
    function AtlasBasePrivateEndpoint(scope, id, props) {
        var _this = _super.call(this, scope, id, props) || this;
        var privEnd = new privateEndpoint.CfnPrivateEndpoint(_this, 'MyProject', {
            groupId: props.projectId,
            region: props.region,
            apiKeys: { privateKey: props.apiKeys.privateKey, publicKey: props.apiKeys.publicKey },
            privateEndpoints: [{ vpcId: props.vpcId, subnetIds: [props.subnetId] }]
        });
        return _this;
    }
    return AtlasBasePrivateEndpoint;
}(cdk.Stack));
var modal = {
    projectId: '',
    vpcId: 'vpc-',
    subnetId: 'subnet-',
    region: 'us-east-1',
    apiKeys: { privateKey: '', publicKey: '' }
};
var app = new aws_cdk_lib_1.App();
var greeter = new AtlasBasePrivateEndpoint(app, 'test-something', modal);
app.synth();
