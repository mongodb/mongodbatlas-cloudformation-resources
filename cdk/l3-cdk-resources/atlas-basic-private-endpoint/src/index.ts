import * as privateEndpoint from '@mongodbatlas-awscdk/private-endpoint';
import * as cdk from 'aws-cdk-lib';
import { App } from 'aws-cdk-lib';
import * as atlas_basic from 'test-atlas-client';

export interface AtlasBasicPropsPrivateEndpoint extends cdk.StackProps{
  readonly projectId : string;
  readonly vpcId : string;
  readonly subnetId : string;
  readonly region : string;
  readonly apiKeys : privateEndpoint.ApiKey;
}

class AtlasBasePrivateEndpoint extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props: AtlasBasicPropsPrivateEndpoint) {
    super(scope, id, props);

    new atlas_basic.AtlasBasic(this, 'QuickStartProject', this.getAtlasBasicInput(props));

    new privateEndpoint.CfnPrivateEndpoint(this, 'QuickStartProject',
      {
        groupId: props.projectId,
        region: props.region,
        apiKeys: { privateKey: props.apiKeys.privateKey, publicKey: props.apiKeys.publicKey },
        privateEndpoints: [{ vpcId: props.vpcId, subnetIds: [props.subnetId] }],
      });
  }

  getAtlasBasicInput(a: AtlasBasicPropsPrivateEndpoint) : atlas_basic.AtlasBasicProps {
    return {
      apiKeys: a.apiKeys,
      clusterProps: {
        apiKeys: a.apiKeys,
        projectId: a.projectId,
        name: 'Cluster_1',
      },
      dbUserProps: {
        databaseName: 'database_1',
        projectId: a.projectId,
        roles: [{
          roleName: 'atlasAdmin',
          databaseName: 'admin',
        }],
        username: 'testUser',
        apiKeys: a.apiKeys,
      },
      ipAccessListProps: {
        apiKeys: a.apiKeys,
        projectId: a.projectId,
        accessList: [
          {
            ipAddress: '0.0.0.0/1',
            comment: 'Testing open all ips',
          },
        ],
      },
      projectId: a.projectId,
    };
  }
}

const modal: AtlasBasicPropsPrivateEndpoint = {
  projectId: '',
  vpcId: 'vpc-',
  subnetId: 'subnet-',
  region: 'us-east-1',
  apiKeys: { privateKey: '', publicKey: '' },
};

const app = new App();

let greeter = new AtlasBasePrivateEndpoint(app, 'test-something', modal );

app.synth();