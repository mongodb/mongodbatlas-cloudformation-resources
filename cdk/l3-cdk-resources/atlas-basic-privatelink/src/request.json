{
    "name": "CDK-L3-Cluster",
    "projectName": "cdk-project",
    "orgId": "5fe4ea50d1a2b617175ee3d4",
    "clusterName": "cdk-cluster",
    "publicKey": "",
    "privateKey": "",
    "clusterType": "REPLICASET",
    "backupEnabled": false,
    "replicationSpecs":[
        {
        "numShards": 1,
        "advancedRegionConfigs": [
            {
                "analyticsSpecs": {
                    "ebsVolumeType": "STANDARD",
                    "instanceSize": "M10",
                    "nodeCount": 1
                  },
                  "electableSpecs": {
                    "ebsVolumeType": "STANDARD",
                    "instanceSize": "M10",
                    "nodeCount": 3
                  },
                  "priority":  7,
                  "regionName": "US_EAST_1"
            }
        ]
    }
    ],
    "databaseName": "admin",
    "roles": [{
      "roleName": "atlasAdmin",
      "databaseName": "admin"
    }],
    "username": "testUser",
    "password": "test123",
    "accessList": [
        {
          "ipAddress": "0.0.0.0/1",
          "comment": "Testing open all ips"
        }
      ]
}