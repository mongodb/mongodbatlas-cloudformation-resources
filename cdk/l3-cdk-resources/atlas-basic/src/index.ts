import * as atlas from '@mongodbatlas-awscdk/cluster';
import * as user from '@mongodbatlas-awscdk/database-user';
import * as project from '@mongodbatlas-awscdk/project';
import * as ipAccessList from '@mongodbatlas-awscdk/project-ip-access-list';
import { Construct } from 'constructs';

/** @type {*} */
const projectDefaults = {
  projectName: 'atlas-project-',
};
/** @type {*} */
const dbDefaults = {
  dbName: 'admin',
  username: 'atlas-user',
  password: 'atlas-pwd',
  roles: [{
    roleName: 'atlasAdmin',
    databaseName: 'admin',
  }],
};
/** @type {*} */
const clusterDefaults = {
  clusterName: 'atlas-cluster-',
  clusterType: 'REPLICASET',
};

/**
 * @description
 * @export
 * @interface ProjectProps
 */
export interface ProjectProps {
  /**
     * @description
     * @type {string}
     * @memberof ProjectProps
     */
  readonly name?: string;
  /**
     * @description
     * @type {string}
     * @memberof ProjectProps
     * @default auto-generated
     */
  readonly orgId: string;
  /**
     * @description
     * @type {string}
     * @memberof ProjectProps
     */
  readonly projectOwnerId?: string;
  /**
     * @description
     * @type {boolean}
     * @memberof ProjectProps
     */
  readonly withDefaultAlertsSettings?: boolean;
  /**
     * @description
     * @type {number}
     * @memberof ProjectProps
     */
  readonly clusterCount?: number;
  /**
     * @description
     * @type {project.ProjectSettings}
     * @memberof ProjectProps
     */
  readonly projectSettings?: project.ProjectSettings;
  /**
     * @description
     * @type {project.ProjectTeam[]}
     * @memberof ProjectProps
     */
  readonly projectTeams?: project.ProjectTeam[];
  /**
     * @description
     * @type {project.ProjectApiKey[]}
     * @memberof ProjectProps
     */
  readonly projectApiKeys?: project.ProjectApiKey[];
}
/**
 * @description
 * @export
 * @interface ClusterProps
 */
export interface ClusterProps {
  /**
     * @description
     * @type {atlas.ProcessArgs}
     * @memberof ClusterProps
     */
  readonly advancedSettings?: atlas.ProcessArgs;
  /**
     * @description
     * @type {boolean}
     * @memberof ClusterProps
     */
  readonly backupEnabled?: boolean;
  /**
     * @description
     * @type {atlas.CfnClusterPropsBiConnector}
     * @memberof ClusterProps
     */
  readonly biConnector?: atlas.CfnClusterPropsBiConnector;
  /**
     * @description
     * @type {string}
     * @memberof ClusterProps
     */
  readonly clusterType?: string;
  /**
     * @description
     * @type {atlas.ConnectionStrings}
     * @memberof ClusterProps
     * @default REPLICASET
     */
  readonly connectionStrings?: atlas.ConnectionStrings;
  /**
     * @description
     * @type {number}
     * @memberof ClusterProps
     */
  readonly diskSizeGb?: number;
  /**
     * @description
     * @type {atlas.CfnClusterPropsEncryptionAtRestProvider}
     * @memberof ClusterProps
     */
  readonly encryptionAtRestProvider?: atlas.CfnClusterPropsEncryptionAtRestProvider;
  /**
     * @description
     * @type {string}
     * @memberof ClusterProps
     */
  readonly projectId?: string;
  /**
     * @description
     * @type {atlas.CfnClusterPropsLabels[]}
     * @memberof ClusterProps
     */
  readonly labels?: atlas.CfnClusterPropsLabels[];
  /**
     * @description
     * @type {string}
     * @memberof ClusterProps
     */
  readonly mongoDbMajorVersion?: string;
  /**
     * @description
     * @type {string}
     * @memberof ClusterProps
     */
  readonly name?: string;
  /**
     * @description
     * @type {boolean}
     * @memberof ClusterProps
     * @default auto-generated
     */
  readonly paused?: boolean;
  /**
     * @description
     * @type {boolean}
     * @memberof ClusterProps
     */
  readonly pitEnabled?: boolean;
  /**
     * @description
     * @type {atlas.AdvancedReplicationSpec[]}
     * @memberof ClusterProps
     */
  readonly replicationSpecs?: atlas.AdvancedReplicationSpec[];
  /**
     * @description
     * @type {string}
     * @memberof ClusterProps
     */
  readonly rootCertType?: string;
  /**
     * @description
     * @type {string}
     * @memberof ClusterProps
     */
  readonly versionReleaseSystem?: string;
  /**
     * @description
     * @type {boolean}
     * @memberof ClusterProps
     */
  readonly terminationProtectionEnabled?: boolean;
}
/**
 * @description
 * @export
 * @interface DatabaseUserProps
 */
export interface DatabaseUserProps {
  /**
     * @description
     * @type {string}
     * @memberof DatabaseUserProps
     */
  readonly deleteAfterDate?: string;
  /**
     * @description
     * @type {user.CfnDatabaseUserPropsAwsiamType}
     * @memberof DatabaseUserProps
     */
  readonly awsiamType?:user.CfnDatabaseUserPropsAwsiamType;
  /**
     * @description
     * @type {string}
     * @memberof DatabaseUserProps
     */
  readonly databaseName?: string;
  /**
     * @description
     * @type {user.LabelDefinition[]}
     * @memberof DatabaseUserProps
     * @default admin
     */
  readonly labels?: user.LabelDefinition[];
  /**
     * @description
     * @type {user.CfnDatabaseUserPropsLdapAuthType}
     * @memberof DatabaseUserProps
     */
  readonly ldapAuthType?: user.CfnDatabaseUserPropsLdapAuthType;
  /**
     * @description
     * @type {user.CfnDatabaseUserPropsX509Type}
     * @memberof DatabaseUserProps
     */
  readonly x509Type?: user.CfnDatabaseUserPropsX509Type;
  /**
     * @description
     * @type {string}
     * @memberof DatabaseUserProps
     */
  readonly password?: string;
  /**
     * @description
     * @type {string}
     * @default cdk-pwd
     * @memberof DatabaseUserProps
     */
  readonly projectId?: string;
  /**
     * @description
     * @type {user.RoleDefinition[]}
     * @memberof DatabaseUserProps
     */
  readonly roles?: user.RoleDefinition[];
  /**
     * @description
     * @type {user.ScopeDefinition[]}
     * @memberof DatabaseUserProps
     */
  readonly scopes?: user.ScopeDefinition[];
  /**
     * @description
     * @type {string}
     * @memberof DatabaseUserProps
     * @default cdk-user
     */
  readonly username?: string;
}
/**
 * @description
 * @export
 * @interface IpAccessListProps
 */
export interface IpAccessListProps {
  /**
     * @description
     * @type {ipAccessList.AccessListDefinition[]}
     * @memberof IpAccessListProps
     */
  readonly accessList: ipAccessList.AccessListDefinition[];
  /**
     * @description
     * @type {string}
     * @memberof IpAccessListProps
     * @default allow-all
     */
  readonly projectId?: string;
  /**
     * @description
     * @type {number}
     * @memberof IpAccessListProps
     */
  readonly totalCount?: number;
  /**
     * @description
     * @type {ipAccessList.ListOptions}
     * @memberof IpAccessListProps
     */
  readonly listOptions?: ipAccessList.ListOptions;
}

/**
 * @description
 * @export
 * @interface AtlasBasicProps
 */
export interface AtlasBasicProps {
  /**
   * @description
   * @type {string}
   * @memberof AtlasBasicProps
   */
  readonly profile ?: string;

  /**
     * @description
     * @type {ProjectProps}
     * @memberof AtlasBasicProps
     */
  readonly projectProps: ProjectProps;
  /**
     * @description
     * @type {ClusterProps}
     * @memberof AtlasBasicProps
     */
  readonly clusterProps: ClusterProps;
  /**
     * @description
     * @type {DatabaseUserProps}
     * @memberof AtlasBasicProps
     */
  readonly dbUserProps?: DatabaseUserProps;
  /**
     * @description
     * @type {IpAccessListProps}
     * @memberof AtlasBasicProps
     */
  readonly ipAccessListProps?: IpAccessListProps;
}
/**
 * @description
 * @export
 * @class AtlasBasic
 * @extends {Construct}
 */
export class AtlasBasic extends Construct {
  /**
   * @description
   * @type {project.CfnProject}
   * @memberof AtlasBasic
   */
  readonly mProject : project.CfnProject;
  /**
   * @description
   * @type {atlas.CfnCluster}
   * @memberof AtlasBasic
   */
  readonly mCluster : atlas.CfnCluster;
  /**
   * @description
   * @type {user.CfnDatabaseUser}
   * @memberof AtlasBasic
   */
  readonly mDBUser: user.CfnDatabaseUser;
  /**
   * @description
   * @type {ipAccessList.CfnProjectIpAccessList}
   * @memberof AtlasBasic
   */
  readonly ipAccessList: ipAccessList.CfnProjectIpAccessList;

  /**
   * Creates an instance of AtlasBasic.
   * @param {Construct} scope
   * @param {string} id
   * @param {AtlasBasicProps} props
   * @memberof AtlasBasic
   */
  constructor(scope: Construct, id: string, props: AtlasBasicProps) {
    super(scope, id);
    //Create a new MongoDB Atlas Project
    this.mProject = new project.CfnProject(this, 'project-'.concat(id), {
      profile: props.profile,
      name: props.projectProps.name || projectDefaults.projectName.concat(String(randomNumber())),
      ...props.projectProps,
    });
    // Create a new MongoDB Atlas Cluster and pass project ID
    this.mCluster = new atlas.CfnCluster(this, 'cluster-'.concat(id),
        {
          profile: props.profile,
          name: props.clusterProps.name || clusterDefaults.clusterName.concat(String(randomNumber())),
          projectId: this.mProject.attrId,
          clusterType: clusterDefaults.clusterType,
          ...props.clusterProps,
        },
    );
    // Create a new MongoDB Atlas Database User
    this.mDBUser = new user.CfnDatabaseUser(this, 'db-user-'.concat(id),
        {
          profile: props.profile,
          databaseName: props.dbUserProps?.databaseName || dbDefaults.dbName,
          projectId: this.mProject.attrId,
          username: props.dbUserProps?.username || dbDefaults.username,
          roles: props.dbUserProps?.roles || dbDefaults.roles,
          password: props.dbUserProps?.password || dbDefaults.password,
          ...props.dbUserProps,
        });
    // Create a new MongoDB Atlas Project IP Access List
    this.ipAccessList = new ipAccessList.CfnProjectIpAccessList(this, 'ip-access-list-'.concat(id),
        {
          profile: props.profile,
          projectId: this.mProject.attrId,
          accessList: props.ipAccessListProps?.accessList,
          ...props.ipAccessListProps,
        },
    );

  }
}

function randomNumber() {
  const min = 10;
  const max = 9999999;
  return Math.floor(Math.random() * (max - min + 1) + min);
}
