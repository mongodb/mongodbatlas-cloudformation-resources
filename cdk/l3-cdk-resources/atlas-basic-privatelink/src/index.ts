import * as atlas from '@mongodbatlas-awscdk/cluster';
import * as user from '@mongodbatlas-awscdk/database-user';
import * as project from '@mongodbatlas-awscdk/project';
import * as ipAccessList from '@mongodbatlas-awscdk/project-ip-access-list';
import { App, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
// import * as privateEndpoint from '@mongodbatlas-awscdk/private-endpoint';
import * as request from './request.json';

export class ClusterL3 extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // Create a new MongoDB Atlas Project
    const myProject = new project.CfnProject(this, 'MyProject', {
      name: request.projectName,
      orgId: request.orgId,
      apiKeys: {
        publicKey: request.publicKey,
        privateKey: request.privateKey,
      },
    });
    // Create a new MongoDB Atlas Cluster and pass project ID
    new atlas.CfnCluster(this, 'MyCluster', {
      apiKeys: {
        publicKey: request.publicKey,
        privateKey: request.privateKey,
      },
      name: request.clusterName,
      projectId: myProject.ref,
      clusterType: request.clusterType,
      backupEnabled: request.backupEnabled,
      replicationSpecs: request.replicationSpecs,
    });

    // Create a new MongoDB Atlas Database User
    new user.CfnDatabaseUser(this, 'MyUser', {
      databaseName: request.databaseName,
      projectId: myProject.ref,
      roles: request.roles,
      username: request.username,
      password: request.password,
      apiKeys: {
        publicKey: request.publicKey,
        privateKey: request.privateKey,
      },
    });

    // Create a new MongoDB Atlas Project IP Access List
    new ipAccessList.CfnProjectIpAccessList(this, 'MyIpAccessList', {
      accessList: request.accessList,
      apiKeys: {
        publicKey: request.publicKey,
        privateKey: request.privateKey,
      },
      projectId: myProject.ref,
    });

    // new privateEndpoint.CfnPrivateEndpoint(this, 'MyPrivateEndpoint',{
    //     groupId: myProject.ref,
    //     region: "us-east-1",
    //     endpointServiceName:"",
    //     errorMessage: "L3 CDK : Private EndPoint creation failed.",
    //     privateEndpoints: [],
    //     status: "",
    //     apiKeys: {
    //       publicKey:"",
    //       privateKey:""
    //     },
    // });
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: '', //process.env.CDK_DEFAULT_ACCOUNT,
  region: '', //process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new ClusterL3(app, request.name, { env: devEnv });

app.synth();