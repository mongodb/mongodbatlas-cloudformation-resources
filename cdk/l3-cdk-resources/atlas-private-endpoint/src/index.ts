// Copyright 2023 MongoDB Inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


import {AtlasBasic, AtlasBasicProps} from "@mongodbatlas-awscdk/atlas-basic";
import {Construct} from "constructs";
import {
    CfnPrivateEndpoint, PrivateEndpoint,
} from "@mongodbatlas-awscdk/private-endpoint";

/** @type {*} */
const privateEndpointDefaults = {
    region: 'us-east-1'
};

/**
 * @description
 * @export
 * @class AtlasBasicPrivateEndpoint
 * @extends {Construct}
 */
export class AtlasBasicPrivateEndpoint extends Construct {

    readonly atlasBasic : AtlasBasic;
    readonly privateEndpoint : CfnPrivateEndpoint;

    /**
     * Creates an instance of AtlasBasicPrivateEndpoint.
     * @param {Construct} scope
     * @param {string} id
     * @param {AtlasBasicPrivateEndpointProps} props
     * @memberof AtlasBasicPrivateEndpoint
     */
    constructor(scope: Construct, id: string, props: AtlasBasicPrivateEndpointProps) {
        super(scope, id);
        validate(props)
        // Validation for atlas basic would be delegated to the library
        this.atlasBasic = new AtlasBasic(this, 'atlas-basic-'.concat(id),
            {
                profile : props.profile,
                ...props.atlasBasicProps,
            });
        this.privateEndpoint = new CfnPrivateEndpoint(this, 'private-endpoint-'.concat(id),
            {
                profile: props.profile,
                groupId: this.atlasBasic.mProject.attrId,
                region: props.region || privateEndpointDefaults.region,
                ...props.privateEndpointProps
            });
    }
}

const validate = (props: AtlasBasicPrivateEndpointProps) => {
    if (!props.atlasBasicProps.projectProps.orgId) {throw Error('Validation error: orgId is not defined');}
};


/**
 * @description
 * @export
 * @interface AtlasBasicPrivateEndpointProps
 */
export interface AtlasBasicPrivateEndpointProps {

    readonly profile?: string;

    /**
     * @description
     * @type {string}
     * @default us-east-1
     * @memberof AtlasPrivateEndpointProps
     */
    readonly region?: string;
    /**
     * @description
     * @type {AtlasBasicProps}
     * @memberof AtlasPrivateEndpointProps
     */
    readonly atlasBasicProps : AtlasBasicProps;
    /**
     * @description
     * @type {CfnPrivateEndpointProps}
     * @memberof AtlasPrivateEndpointProps
     */
    readonly privateEndpointProps: PrivateEndpointProps;
}

/**
 * @description
 * @export
 * @interface PrivateEndpointProps
 */
export interface PrivateEndpointProps {
    /**
     * @description
     * @type {string}
     * @memberof PrivateEndpointProps
     */
    readonly endpointServiceName?: string;
    /**
     * @description
     * @type {string}
     * @memberof PrivateEndpointProps
     */
    readonly errorMessage?: string;
    /**
     * @description
     * @type {string}
     * @memberof PrivateEndpointProps
     */
    readonly status?: string;
    /**
     * @description
     * @type {string}
     * @memberof PrivateEndpointProps
     */
    readonly groupId?: string;
    /**
     * @description
     * @type {string}
     * @memberof PrivateEndpointProps
     */
    readonly region?: string;
    /**
     * @description
     * @type {PrivateEndpoint[]}
     * @memberof PrivateEndpointProps
     */
    readonly privateEndpoints?: PrivateEndpoint[];

}
