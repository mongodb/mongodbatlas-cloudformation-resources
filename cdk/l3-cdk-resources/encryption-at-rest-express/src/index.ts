import * as atlas from '@mongodbatlas-awscdk/atlas-basic';
import * as cluster from '@mongodbatlas-awscdk/cluster';
import * as databaseUser from '@mongodbatlas-awscdk/database-user';
import * as encryption from '@mongodbatlas-awscdk/encryption-at-rest';
import * as accessList from '@mongodbatlas-awscdk/project-ip-access-list';
import {Construct} from 'constructs';


const NODE_COUNT = 3;
const NODE_COUNT_ANALYTICS = 1;
const BACKUP_ENABLED = true;
const INSTANCE_SIZE = 'M30';
const MONGODB_VERSION = '5.0';
const ENCRYPTION_AT_REST_PROVIDER = cluster.CfnClusterPropsEncryptionAtRestProvider.AWS;
const REGION = 'US_EAST_1';
const EBS_VOLUME_TYPE = 'STANDARD';
const ENABLE_ENCRYPTION_AT_REST = true;
const CLUSTER_TYPE = 'REPLICASET';
const DB_NAME = 'admin';
const USERNAME = 'cdkUser';
const ROLE = [{
  roleName: 'atlasAdmin',
  databaseName: 'admin',
}];


function randomNumber() {
  const min = 10;
  const max = 9999999;
  return Math.floor(Math.random() * (max - min + 1) + min);
}

function getClusterProps(inputClusterProps: atlas.ClusterProps): cluster.CfnClusterProps {
  return {
    name: inputClusterProps.name || 'atlas-cluster-'.concat(String(randomNumber())),
    mongoDbMajorVersion: inputClusterProps.mongoDbMajorVersion || MONGODB_VERSION,
    backupEnabled: inputClusterProps.backupEnabled || BACKUP_ENABLED,
    diskSizeGb: inputClusterProps.diskSizeGb,
    clusterType: inputClusterProps.clusterType || CLUSTER_TYPE,
    biConnector: inputClusterProps.biConnector,
    connectionStrings: inputClusterProps.connectionStrings,
    encryptionAtRestProvider: inputClusterProps.encryptionAtRestProvider || ENCRYPTION_AT_REST_PROVIDER,
    labels: inputClusterProps.labels,
    paused: inputClusterProps.paused,
    pitEnabled: inputClusterProps.pitEnabled,
    rootCertType: inputClusterProps.rootCertType,
    terminationProtectionEnabled: inputClusterProps.terminationProtectionEnabled,
    versionReleaseSystem: inputClusterProps.versionReleaseSystem,
    advancedSettings: inputClusterProps.advancedSettings,
    replicationSpecs: inputClusterProps.replicationSpecs || getDefaultClusterReplicationSpec(),
    projectId: '',
  };
}

function getDefaultClusterReplicationSpec(): cluster.AdvancedReplicationSpec[] {
  return [
    {
      numShards: 1,
      advancedRegionConfigs: [
        {
          analyticsSpecs: {
            ebsVolumeType: EBS_VOLUME_TYPE,
            instanceSize: INSTANCE_SIZE,
            nodeCount: NODE_COUNT_ANALYTICS,
          },
          electableSpecs: {
            ebsVolumeType: EBS_VOLUME_TYPE,
            instanceSize: INSTANCE_SIZE,
            nodeCount: NODE_COUNT,
          },
          priority: 7,
          regionName: REGION,
        },
      ],
    },
  ];

}

export interface AtlasEncryptionAtRestProps{
  /**
       * ID of an AWS IAM role authorized to manage an AWS customer master key.
       *
       * @schema AwsKms#RoleID
    */
  readonly roleId: string;
  /**
      * The AWS customer master key used to encrypt and decrypt the MongoDB master keys.
      *
      * @schema AwsKms#CustomerMasterKeyID
      */
  readonly customerMasterKeyId: string;
  /**
      * Specifies whether Encryption at Rest is enabled for an Atlas project. To disable Encryption at Rest, pass only this parameter with a value of false. When you disable Encryption at Rest, Atlas also removes the configuration details.
      * Default Value: true
      * @schema AwsKms#Enabled
      */
  readonly enabledEncryptionAtRest?: boolean;

  /**
     * The AWS region in which the AWS customer master key exists.
     *
     * @schema AwsKms#Region
     */
  readonly region?: string;
}

export interface AtlasEncryptionAtRestExpressProps {
  readonly cluster?: atlas.ClusterProps;
  readonly accessList?: atlas.IpAccessListProps;
  readonly encryptionAtRest: AtlasEncryptionAtRestProps;
  readonly databaseUser?: atlas.DatabaseUserProps;
  readonly profile?: string;
  readonly projectId: string;
}

export class AtlasEncryptionAtRestExpress extends Construct {
  readonly encryptionAtRest: encryption.CfnEncryptionAtRest;
  readonly cluster?: cluster.CfnCluster;
  readonly accessList?: accessList.CfnProjectIpAccessList;
  readonly databaseUser?: databaseUser.CfnDatabaseUser;

  constructor(scope: Construct, id: string, props: AtlasEncryptionAtRestExpressProps) {
    super(scope, id);

    this.encryptionAtRest = new encryption.CfnEncryptionAtRest(this, 'encryption-at-rest-'.concat(id), {
      awsKms: {
        customerMasterKeyId: props.encryptionAtRest.customerMasterKeyId,
        roleId: props.encryptionAtRest.roleId,
        enabled: props.encryptionAtRest.enabledEncryptionAtRest || ENABLE_ENCRYPTION_AT_REST,
        region: props.encryptionAtRest.region || REGION,
      },
      projectId: props.projectId,
      profile: props.profile,
    });

    if (props.cluster) {
      // Create a new MongoDB Atlas Cluster and pass project ID
      const clusterProps = getClusterProps(props.cluster);
      this.cluster = new cluster.CfnCluster(this, 'cluster-'.concat(id),
        {
          ...clusterProps,
          profile: props.profile,
          projectId: props.projectId,
        });
    }

    if (props.databaseUser) {
      // Create a new MongoDB Atlas Database User
      this.databaseUser = new databaseUser.CfnDatabaseUser(this, 'db-user-'.concat(id),
        {
          ...props.databaseUser,
          profile: props.profile,
          databaseName: props.databaseUser?.databaseName || DB_NAME,
          projectId: props.projectId,
          username: props.databaseUser?.username || USERNAME,
          roles: props.databaseUser?.roles || ROLE,
          password: props.databaseUser.password,
        });
    }


    if (props.accessList) {
      // Create a new MongoDB Atlas Project IP Access List
      this.accessList = new accessList.CfnProjectIpAccessList(this, 'access-list-'.concat(id),
        {
          ...props.accessList,
          profile: props.profile,
          projectId: props.projectId,
        });
    }
  }
}


