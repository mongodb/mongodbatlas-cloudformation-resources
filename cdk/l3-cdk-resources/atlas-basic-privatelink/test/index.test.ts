import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ClusterL3 } from '../src/index';

test('Snapshot', () => {
  const app = new App();
  const stack = new ClusterL3(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});