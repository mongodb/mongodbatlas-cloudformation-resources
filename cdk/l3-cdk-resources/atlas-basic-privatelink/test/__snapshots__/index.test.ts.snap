// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "MyCluster": Object {
      "Properties": Object {
        "ApiKeys": Object {
          "PrivateKey": "",
          "PublicKey": "",
        },
        "BackupEnabled": false,
        "ClusterType": "REPLICASET",
        "Name": "cdk-cluster",
        "ProjectId": Object {
          "Ref": "MyProject",
        },
        "ReplicationSpecs": Array [
          Object {
            "AdvancedRegionConfigs": Array [
              Object {
                "AnalyticsSpecs": Object {
                  "EbsVolumeType": "STANDARD",
                  "InstanceSize": "M10",
                  "NodeCount": 1,
                },
                "ElectableSpecs": Object {
                  "EbsVolumeType": "STANDARD",
                  "InstanceSize": "M10",
                  "NodeCount": 3,
                },
                "Priority": 7,
                "RegionName": "US_EAST_1",
              },
            ],
            "NumShards": 1,
          },
        ],
      },
      "Type": "MongoDB::Atlas::Cluster",
    },
    "MyIpAccessList": Object {
      "Properties": Object {
        "AccessList": Array [
          Object {
            "Comment": "Testing open all ips",
            "IPAddress": "0.0.0.0/1",
          },
        ],
        "ApiKeys": Object {
          "PrivateKey": "",
          "PublicKey": "",
        },
        "ProjectId": Object {
          "Ref": "MyProject",
        },
      },
      "Type": "MongoDB::Atlas::ProjectIpAccessList",
    },
    "MyProject": Object {
      "Properties": Object {
        "ApiKeys": Object {
          "PrivateKey": "",
          "PublicKey": "",
        },
        "Name": "cdk-project",
        "OrgId": "5fe4ea50d1a2b617175ee3d4",
      },
      "Type": "MongoDB::Atlas::Project",
    },
    "MyUser": Object {
      "Properties": Object {
        "ApiKeys": Object {
          "PrivateKey": "",
          "PublicKey": "",
        },
        "DatabaseName": "admin",
        "Password": "test123",
        "ProjectId": Object {
          "Ref": "MyProject",
        },
        "Roles": Array [
          Object {
            "DatabaseName": "admin",
            "RoleName": "atlasAdmin",
          },
        ],
        "Username": "testUser",
      },
      "Type": "MongoDB::Atlas::DatabaseUser",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
